### PR TITLE
Fix nil pointer dereference in error handling of stats client call to Envoy admin

### PR DESCRIPTION
**Pull request by karanvasnani**
### Summary
In `snapshotter.go` when the call to get stats from Envoy fails (say due to timeout), the else block attempts to dereference `statsResponse.status` when the `statsResponse` is `nil`.
 
This PR fixes the error handling logic in `snapshotter.go` and `envoy_prometheus_stats.go` based on the reading of https://pkg.go.dev/net/http#Client.Do. 

### Testing
Tested by building a custom image with the changes and deploying in my account.

### Description for the changelog
Fix nil pointer dereference in error handling of stats client call to Envoy admin

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/stats/snapshotter.go [hunk 1]
```diff
@@ -247,22 +247,18 @@ func getStatsFromEnvoy(httpClient *retryablehttp.Client, request *retryablehttp.
...
 
 	if err != nil {
-		// Valid status code range is [100, 600)
-		if statsResponse != nil && statsResponse.StatusCode < 600 && statsResponse.StatusCode >= 100 {
-			return nil, &GetStatsError{statsResponse.StatusCode, fmt.Errorf("unable to reach Envoy Admin: %v", err)}
-		} else {
-			return nil, &GetStatsError{http.StatusInternalServerError, fmt.Errorf("unable to reach Envoy Admin, %v. Status: %v, error code: %v", err, statsResponse.Status, statsResponse.StatusCode)}
-		}
-	}
-	if statsResponse == nil {
```
Comment (karanvasnani) [inline agent/stats/snapshotter.go:257]: This block is redundant as the response will always be non-nil if error is nil.

File: agent/stats/snapshotter.go [hunk 2]
```diff
@@ -247,22 +247,18 @@ func getStatsFromEnvoy(httpClient *retryablehttp.Client, request *retryablehttp.
 	log.Debugf("Stats request took: %vms", duration.Milliseconds())
 
 	if err != nil {
-		// Valid status code range is [100, 600)
-		if statsResponse != nil && statsResponse.StatusCode < 600 && statsResponse.StatusCode >= 100 {
```
Comment (karanvasnani) [inline agent/stats/snapshotter.go:251]: This isn't needed as if the error is non-nil, we can ignore the response and just return an error.

File: agent/stats/envoy_prometheus_stats.go [hunk 3]
```diff
@@ -129,32 +120,22 @@ func (envoyPrometheusStatsHandler *EnvoyPrometheusStatsHandler) HandleStats(resW
...
-		log.Debug("empty response from Envoy stat endpoint")
+		log.Errorf("Call to fetch stats from Envoy admin failed: %s", err)
+		http.Error(resWriter, "Failed to fetch stats from Envoy", http.StatusInternalServerError)
 		return
 	}
 
 	if statsResponse.StatusCode != http.StatusOK {
-		http.Error(resWriter, fmt.Sprintf("Envoy response not OK, error code: %v", statsResponse.StatusCode), statsResponse.StatusCode)
+		log.Errorf("Envoy stats response status code not OK: %v", statsResponse.Status)
+		http.Error(resWriter, "Failed to fetch stats from Envoy", http.StatusInternalServerError)
```
Comment (karanvasnani) [inline agent/stats/envoy_prometheus_stats.go:130]: Responding with InternalServerError in all failure cases as, we don't need to leak internal details of interaction with Envoy admin to the client (ecs agent).

File: agent/client/agent_http_client.go [hunk 4]
```diff
@@ -142,10 +142,10 @@ func CreateStandardAgentHttpRequest(method, requestUrl string, body io.Reader) (
 }
 
 func RetryErrorHandler(resp *http.Response, err error, attempt int) (retryResponse *http.Response, retryError error) {
-	if resp != nil && resp.StatusCode < 600 && resp.StatusCode >= 100 {
-		retryError = fmt.Errorf("giving up after %d attempt(s), error: %v, status: %v", attempt, err, resp.Status)
-	} else {
+	if err != nil {
 		retryError = fmt.Errorf("giving up after %d attempt(s), error: %v", attempt, err)
+	} else {
+		retryError = fmt.Errorf("giving up after %d attempt(s), error: %v, status: %v", attempt, err, resp.Status)
```
[the remark this answers is not in the document]
Comment (karanvasnani) [reply agent/client/agent_http_client.go]: Yeah my bad, updated now. Only one of `resp` or `err` will be non-nil and we will use that to return as the error message. Also, added a line to close the `resp.Body` if response is non-nil.

**Review by suniltheta — APPROVED** [review]
just there is a merge conflict that needs to be fixed

**Review by suniltheta — APPROVED** [review]

File: agent/stats/envoy_prometheus_stats.go (hunk 3 above)
Comment (ytsssun) [reply agent/stats/envoy_prometheus_stats.go:130]: I am fine with the change.

Just posting some background here. In the early discussion when we implement this, we wanted to "pass the original return code through to the client. Save the 500s for when the Agent itself has an unrecoverable error".

The clients may not just be ECS agent, if the customer start using SC agent stats directly and not using the Admin interface, they may see different response from the stats endpoints (Envoy vs SC agent).

File: agent/client/agent_http_client.go [hunk 5]
```diff
@@ -142,11 +142,19 @@ func CreateStandardAgentHttpRequest(method, requestUrl string, body io.Reader) (
 }
 
 func RetryErrorHandler(resp *http.Response, err error, attempt int) (retryResponse *http.Response, retryError error) {
-	if resp != nil && resp.StatusCode < 600 && resp.StatusCode >= 100 {
```
Comment (ytsssun) [inline agent/client/agent_http_client.go:145]: I think it is still worth it keeping the validation for statusCode?

**Review by ytsssun — APPROVED** [review]

File: agent/client/agent_http_client.go [hunk 6]
```diff
@@ -142,11 +142,19 @@ func CreateStandardAgentHttpRequest(method, requestUrl string, body io.Reader) (
 }
 
 func RetryErrorHandler(resp *http.Response, err error, attempt int) (retryResponse *http.Response, retryError error) {
```
[the remark this answers is not in the document]
Comment (karanvasnani) [reply agent/client/agent_http_client.go]: Yeah good catch. Changed now

File: agent/client/agent_http_client.go (hunk 5 above)
Comment (karanvasnani) [reply agent/client/agent_http_client.go:145]: Given that is the only possible range of status codes, I believe it is redundant to check and add validations for it

File: agent/stats/envoy_prometheus_stats.go (hunk 3 above)
Comment (karanvasnani) [reply agent/stats/envoy_prometheus_stats.go:130]: I think the idea is to abstract the Envoy interface behind the SC agent's management interface and the client doesn't need to be aware of the internals. I know it is possible currently to directly query Envoy but, that is not expected by the client and may change in future.

**Pull request merged**